### ISSUE TITLE
Language injections: syntax highlight HTML/CSS/JSON... in the surrounding content of the template 

### ISSUE THREAD

**Issue by uncenter** [uncenter/tree-sitter-tera#2]
Detect the language of the surrounding template, and then inject it for `content` rules.

**References**:
- https://github.com/IndianBoy42/tree-sitter-just/blob/6aee3d2f5b3a47286bae0916daaa6c7a217f6fa4/queries/just/injections.scm

**Comment by uncenter** [comment]
Could use `@injection.filename` for Whiskers templates on the `whiskers.filename` property: https://neovim.io/doc/user/treesitter.html#treesitter-language-injections.

Still planning on using some sort of comment syntax as well though, e.g. `{# tera-content-language: toml #}`.

**Comment by uncenter** [comment]
Actually doesn't quite close this. Plan is now to simply base on extensions. Provide multiple Tera "languages", one for each of a set of predefined supported content languages. Implemented in https://github.com/uncenter/zed-tera already.

**Comment by MuntasirSZN** [comment]
I am been a bit with vimscript, added syntax etc to vim upstream. In the code, I am getting the filename then getting the part after first dot and before second dot, and enabling syntax for that. like:

file.toml.tera

It will enable toml syntax highlighting, with tera (tera has more priority). Can we do something similar?

**Comment by uncenter** [comment]
Yes that is the idea. That's how https://github.com/uncenter/zed-tera does it. If you know of a good way to do that in Neovim and Vim, that would be appreciated. Let's talk it out first before we do anything upstream though.

**Comment by MuntasirSZN** [comment]
> Yes that is the idea. That's how https://github.com/uncenter/zed-tera does it. If you know of a good way to do that in Neovim and Vim, that would be appreciated. Let's talk it out first before we do anything upstream though.

Ok. Though what I did won't hit this at all, its just all vim related. We can get help from clason also.

**Comment by MuntasirSZN** [comment]
We should try to use treesitter, not lua or vimscript.

**Comment by tombh** [comment]
I'm still a bit new to Treesitter inner workings. So this issue is about being able to highlight both, say HTML, and Tera in a single file? And there's currently no way to do that? (at least in nvim where I am?)

**Comment by uncenter** [comment]
No, it is possible, the issue is that Tera is used for more than just HTML templating and it is difficult to make that configurable/work for everyone. You could make HTML the default injected language but then it wouldn't work for people who use Tera to template CSS, for example.

**Comment by uncenter** [comment]
As @MuntasirSZN mentioned above, one way to allow for configuration would be based on the filename; `foo.css.tera` has CSS highlights, `bar.toml.tera` has TOML highlights, `baz.html.tera` (and maybe `baz.tera` if we make it the default) has HTML highlights. I'm not familiar enough with Lua (nor do I use Neovim regularly) to make this work, but I'm happy to do whatever I can on my end to make that work.

**Comment by MuntasirSZN** [comment]
There is another way. Comments.

**Comment by uncenter** [comment]
> There is another way. Comments.

This is not possible with Tree-sitter since one node cannot change injections for all other `content` nodes in the tree.

**Comment by MuntasirSZN** [comment]
> > There is another way. Comments.
> 
> This is not possible with Tree-sitter since one node cannot change injections for all other `content` nodes in the tree.

Hmmm.

**Comment by tombh** [comment]
> Tera is used for more than just HTML templating

Ohhh, right, that makes total sense now. Thank you.

So in theory could I build my own personal version using the closed PR from here? https://github.com/uncenter/tree-sitter-tera/pull/10

**Comment by uncenter** [comment]
Yep! This is the line you would want to add to your *Tree-sitter injections*.

```scm
((content) @injection.content
  (#set! injection.language "html")
  (#set! injection.combined))
```

You may already know this but it seems like you can create a custom injections file like so: https://github.com/gmr458/nvim/blob/main/queries/typescript/injections.scm?
Haven't tested this but it might look something like:

```scm
; extends

((content) @injection.content
  (#set! injection.language "html")
  (#set! injection.combined))
```

In `queries/tera/injections.scm` of your Neovim configuration?

**Comment by tombh** [comment]
Wow, thank you 🥹 If I get it working I'll report back!

**Comment by uncenter** [comment]
No problem. I'll look into making some kind of Lua plugin to do that automatically and depending on the file extension :)

**Comment by MuntasirSZN** [comment]
> No problem. I'll look into making some kind of Lua plugin to do that automatically and depending on the file extension :)

I can do it.

**Comment by MuntasirSZN** [comment]
See this https://github.com/MuntasirSZN/tera-autoextsyn.nvim (Couldn't find a name).

**Comment by nabiu256** [comment]
>Haven't tested this but it might look something like:
> 
>
```scm
 ; extends
 
 ((content) @injection.content
   (#set! injection.language "html")
   (#set! injection.combined))
```
> 
> In `queries/tera/injections.scm` of your Neovim configuration?

After hours of trying to get proper Tera + HTML syntax highlighting in Neovim I found this Issue and I can confirm that, as far as I can tell, this does work. It seems break the `rainbow-delimiters` plugin which makes all tags have the same color but, other than that, now the parsing is correct and I can get proper highlights and folds.

**Comment by uncenter** [comment]
> After hours of trying to get proper Tera + HTML syntax highlighting in Neovim I found this Issue

I'm so sorry you had to spend so much time on this! I'd really like to make the default HTML highlighting but my primary usecase of Tera isn't even HTML templating and it's apparently quite difficult to reconcile the two priorities. I like the simplicity of the approach I landed on for users in https://github.com/uncenter/zed-tera?tab=readme-ov-file#default-to-html-highlighting-of-content (though it isn't the greatest setup in terms of how it functions under the hood...). Helix and Neovim don't seem to offer as much configuration capability since the syntax highlighting in those two cases is built in rather than from an extension. Neovim has a plugin ecosystem, and there was a little discussion above about how it would work with that, but I don't use Neovim and am not familiar with Lua to make that possible. I would love to work with someone who is interested in solving the problem and has experience with Neovim/Lua if I can be of any assistance on the Tree-sitter and Tera sides of things!

> I can confirm that, as far as I can tell, this does work

I guess in the mean time I'd be happy to add this as a documentation tip to the README of this project. However if it took you a while to even find this repository, maybe it's best to start with defaulting to HTML highlighting - likely the majority of users' usecases - and work backwards from there to support my own?
